rfq/portfolio_pilot.go

```diff
@@ -0,0 +1,166 @@
+	r.outcome.WhenLeft(pred)
+}
+
+// WhenReject executes the callback with the rejection error when the response
+// is a rejection and does nothing otherwise.
+func (r *ResolveResp) WhenReject(pred func(err rfqmsg.RejectErr)) {
+	r.outcome.WhenRight(pred)
+}
+
+// PortfolioPilot evaluates RFQs and returns either an accepted asset rate quote
+// or a rejection reason.
+type PortfolioPilot interface {
+	// ResolveRequest resolves a quote request by returning either an
+	// acceptable asset rate or a quote rejection error. Errors are reserved
+	// for unexpected failures while evaluating the request.
+	ResolveRequest(context.Context, rfqmsg.Request) (ResolveResp,
+		error)
+}
+
+// InternalPortfolioPilotConfig holds settings for the built-in pilot that uses
+// a price oracle for pricing decisions.
+type InternalPortfolioPilotConfig struct {
+	// PriceOracle supplies pricing data. If nil, the pilot rejects requests
+	// as the oracle is considered unavailable.
+	PriceOracle PriceOracle
+
+	// ForwardPeerIDToOracle controls whether the requesting peer ID is sent
+	// to the oracle for peer-specific pricing. Disabling this avoids
+	// sharing caller identity with the oracle.
+	ForwardPeerIDToOracle bool
+}
+
+// InternalPortfolioPilot is the built-in RFQ decision logic that delegates
+// pricing to an external price oracle.
+type InternalPortfolioPilot struct {
+	// cfg holds settings and supporting components for the portfolio pilot.
+	cfg InternalPortfolioPilotConfig
+}
+
+// NewInternalPortfolioPilot constructs a new pilot using the provided config.
+func NewInternalPortfolioPilot(
+	cfg InternalPortfolioPilotConfig) *InternalPortfolioPilot {
+
+	return &InternalPortfolioPilot{
+		cfg: cfg,
+	}
+}
+
+// ResolveRequest resolves a quote request by querying the configured price
+// oracle. It accepts with a quote when a valid rate is returned, rejects when
+// no oracle is configured, and errors on unexpected oracle failures or missing
+// rates.
+func (p *InternalPortfolioPilot) ResolveRequest(ctx context.Context,
+	request rfqmsg.Request) (ResolveResp, error) {
+
+	var zero ResolveResp
+
+	if p.cfg.PriceOracle == nil {
+		return NewRejectResolveResp(
+			rfqmsg.ErrPriceOracleUnavailable,
+		), nil
+	}
+
+	peerID := fn.None[route.Vertex]()
+	if p.cfg.ForwardPeerIDToOracle {
+		peerID = fn.Some(request.MsgPeer())
+	}
+
+	var resp *OracleResponse
+
+	switch req := request.(type) {
+	case *rfqmsg.BuyRequest:
+		var err error
+		resp, err = p.cfg.PriceOracle.QuerySellPrice(
+			ctx, req.AssetSpecifier, fn.Some(req.AssetMaxAmt),
+			fn.None[lnwire.MilliSatoshi](), req.AssetRateHint,
+			peerID, req.PriceOracleMetadata, IntentRecvPayment,
+		)
+		if err != nil {
+			return zero, fmt.Errorf("query sell price: %w", err)
+		}
+
+	case *rfqmsg.SellRequest:
+		var err error
+		resp, err = p.cfg.PriceOracle.QueryBuyPrice(
+			ctx, req.AssetSpecifier, fn.None[uint64](),
+			fn.Some(req.PaymentMaxAmt), fn.None[rfqmsg.AssetRate](),
```

Comment:
Intentional that `request.AssetRateHint` is not passed here anymore?

Reply:
I will pass in.